### DISCUSSION TITLE
How to add artifacts from another image to ddev web-build?

### DISCUSSION THREAD

**Discussion by MurzNN** [ddev/ddev#5581] [Q&A]
I need to build a custom utility, using another image, and copy the results to the ddev web image.
So, I created a separate Dockerfile.xk6-elastic with the content:
```
FROM golang:1.20-alpine AS k6-build
WORKDIR $GOPATH/src/go.k6.io/k6
ADD . .
RUN apk --no-cache add git
RUN CGO_ENABLED=0 go install go.k6.io/xk6/cmd/xk6@latest
RUN CGO_ENABLED=0 xk6 build --with github.com/elastic/xk6-output-elasticsearch@v0.2.0 --output /tmp/k6

# Adding "arg base-image"
ARG BASE_IMAGE
# Using "from as" construction
FROM $BASE_IMAGE AS ${BASE_IMAGE}_final
# Using "copy" construction
COPY --from=k6-build /tmp/k6 /usr/bin/k6
# End k6
```
But after restarting the ddev - I got the error:
```
failed to solve: circular dependency detected on stage: k6-build
```
How can I resolve this problem?

**Comment by MurzNN** [comment]
Seems this happens because `ddev` removes the line:
```
FROM $BASE_IMAGE AS ${BASE_IMAGE}_final
``` 
from the final compiled Dockerfile, here is the contents of the `.ddev/.webimageBuild/Dockerfile`:
```
### From user Dockerfile /home/alexey_korepov/projects/myproject/.ddev/web-build/Dockerfile.k6-elastic:
FROM golang:1.20-alpine AS k6-build
WORKDIR $GOPATH/src/go.k6.io/k6
ADD . .
RUN apk --no-cache add git
RUN CGO_ENABLED=0 go install go.k6.io/xk6/cmd/xk6@latest
RUN CGO_ENABLED=0 xk6 build --with github.com/elastic/xk6-output-elasticsearch@v0.2.0 --output /tmp/k6

# Adding "arg base-image"
# Using "from as" construction
# Using "copy" construction
COPY --from=k6-build /tmp/k6 /usr/bin/k6
# End k6
```
So the line just disappeared!

**Reply by stasadev** [reply]
Hi @MurzNN,

You probably need to use a `.ddev/web-build/pre.Dockerfile.*`

https://ddev.readthedocs.io/en/latest/users/extend/customizing-images/?h=pre.#adding-extra-dockerfiles-for-webimage-and-dbimage

Try to rename your custom `Dockerfile.xk6-elastic` to `pre.Dockerfile.xk6-elastic`.

If anyone is wondering why `FROM $BASE_IMAGE` was removed there, see https://github.com/ddev/ddev/pull/3999

**Reply by MurzNN** [reply]
Thanks, it partly works, but the "pre" file is inserted after some commands already executed, so seems I should "replay" all the commands again, so here is the working version of the `pre.Dockerfile.xk6-elastic` file:
```
FROM golang:1.20-alpine AS k6-build
WORKDIR $GOPATH/src/go.k6.io/k6
ADD . .
RUN apk --no-cache add git
RUN CGO_ENABLED=0 go install go.k6.io/xk6/cmd/xk6@latest
RUN CGO_ENABLED=0 xk6 build --with github.com/elastic/xk6-output-elasticsearch@v0.2.0 --output /tmp/k6

ARG BASE_IMAGE
FROM $BASE_IMAGE
COPY --from=k6-build /tmp/k6 /usr/bin/k6

# Copy of the commands from the ddev's initial "FROM"
SHELL ["/bin/bash", "-c"]

ARG username
ARG uid
ARG gid
ARG DDEV_PHP_VERSION
RUN (groupadd --gid $gid "$username" || groupadd "$username" || true) && (useradd  -l -m -s "/bin/bash" --gid "$username" --comment '' --uid $uid "$username" || useradd  -l -m -s "/bin/bash" --gid "$username" --comment '' "$username" || useradd  -l -m -s "/bin/bash" --gid "$gid" --comment '' "$username" || useradd -l -m -s "/bin/bash" --comment '' $username )
```
Is there a way to make it works without copy-pasting the ddev's commands?

**Reply by rfay** [reply]
(Just a note, you probably already know this but the concatenated Dockerfile is built into .ddev/.webimageBuild/Dockerfile, so you can study it there. That's probably what you're already doing.)

What you're doing is pretty exotic :)

**Reply by MurzNN** [reply]
@rfay Not so exotic from my point of view ;)  I just need to compile a `k6` utility to use inside ddev - here is their original  Dockerfile https://github.com/elastic/xk6-output-elasticsearch/blob/main/Dockerfile

Maybe you could advise a more easy way to compile a tool in Golang, to integrate it into the `ddev-web` container?

**Reply by stasadev** [reply]
I don't use such builds, so I can't recommend something better. If this `k6` is used in many DDEV projects, it might be worth building it somewhere else and then adding it to the container via curl/wget.

**Reply by rfay** [reply]
You can certainly compile in a Dockerfile; Just install golang or use related technique in the Dockerfile. But @stasadev 's suggestion is probably better. There are several add-ons that build things like PHP extensions, which is harder.

**Reply by MurzNN** [reply]
> You can certainly compile in a Dockerfile; Just install golang or use related technique in the Dockerfile. 

Yes, but this way will keep a lot of garbage after compiling, but I need only the single compiled file as an artifact, with no other files. This is because the approach of using a separate container to compile is much better.

> If this k6 is used in many DDEV projects, it might be worth building it somewhere else and then adding it to the container via curl/wget.

Yes, downloading the built binary is much easier, but it is not provided by the maintainers. And while downloading, I need to also match the correct architecture (linux-amd64, macos-arm64, etc), but with compilation - it compiles to the current architecture automatically.

So my described approach seems to work well, only duplicating the actions looks not so good.

**Reply by rfay** [reply]
> Yes, but this way will keep a lot of garbage after compiling, but I need only the single compiled file as an artifact, with no other files. This is because the approach of using a separate container to compile is much better.

That's actually not true; there are many docker build examples where the intended target is kept and nothing else. 

Anyway, glad you have a solution!
